Objects/dictobject.c

```diff
@@ -7313,20 +7327,67 @@ _PyDictKeys_DecRef(PyDictKeysObject *keys)
     dictkeys_decref(interp, keys, false);
 }
 
-uint32_t _PyDictKeys_GetVersionForCurrentState(PyInterpreterState *interp,
-                                               PyDictKeysObject *dictkeys)
+static inline uint32_t
+get_next_dict_keys_version(PyInterpreterState *interp)
 {
-    if (dictkeys->dk_version != 0) {
-        return dictkeys->dk_version;
-    }
+#ifdef Py_GIL_DISABLED
+    uint32_t v;
+    do {
+        v = _Py_atomic_load_uint32_relaxed(
+            &interp->dict_state.next_keys_version);
+        if (v == 0) {
+            return 0;
+        }
+    } while (!_Py_atomic_compare_exchange_uint32(
+        &interp->dict_state.next_keys_version, &v, v + 1));
+#else
     if (interp->dict_state.next_keys_version == 0) {
         return 0;
     }
     uint32_t v = interp->dict_state.next_keys_version++;
-    dictkeys->dk_version = v;
+#endif
     return v;
 }
 
+typedef struct {
+    uint32_t version;
+    int assigned;
+} assign_keys_version_result;
+
+static inline assign_keys_version_result
+assign_keys_version(PyInterpreterState *interp, PyDictKeysObject *dictkeys)
+{
+    uint32_t dk_version = FT_ATOMIC_LOAD_UINT32_RELAXED(dictkeys->dk_version);
+    if (dk_version != 0) {
+        return (assign_keys_version_result){dk_version, 0};
+    }
+    dk_version = get_next_dict_keys_version(interp);
+    FT_ATOMIC_STORE_UINT32_RELAXED(dictkeys->dk_version, dk_version);
+    return (assign_keys_version_result){dk_version, 1};
+}
+
+// In free-threaded builds the caller must ensure that the keys object is not
+// being mutated concurrently by another thread.
+uint32_t
+_PyDictKeys_GetVersionForCurrentState(PyInterpreterState *interp,
+                                      PyDictKeysObject *dictkeys)
+{
+    assign_keys_version_result res = assign_keys_version(interp, dictkeys);
+    return res.version;
+}
+
+uint32_t
+_PyDict_GetKeysVersionForCurrentState(PyInterpreterState *interp,
+                                      PyDictObject *dict)
+{
+    ASSERT_DICT_LOCKED((PyObject *) dict);
+    assign_keys_version_result res = assign_keys_version(interp, dict->ma_keys);
+    if (res.assigned) {
```

Comment:
I don't think you need the `res.assigned` check and the combination with `_Py_IsOwnedByCurrentThread` in `ensure_shared_on_keys_version_assignment` is not correct, I think: we could end up not marking the dict as shared in a non-owning thread.

Marking the dict as shared is cheap and this is not called frequently, and I expect the code will be simpler without the `assign_keys_version_result`.